### PR TITLE
[clang] Substitute alias templates from correct context

### LINKED ISSUE #41693: Invalid private member diagnostic of template using with decltype of private member method
|  |  |
| --- | --- |
| Bugzilla Link | [42348](https://llvm.org/bz42348) |
| Version | trunk |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@smilingthax,@zygoloid |

## Extended Description 
The following code fails to compile with an erroneous error "'Impl' is a private member of 'S<void>'":

```cpp
  template <typename> struct S
    {
    private:
    template <typename> static constexpr auto Impl();

    public:
    template <typename X> using U = decltype(Impl<X>());
    };

  template <typename T> template <typename>
  constexpr auto S<T>::Impl() { }

  using X = S<void>::U<void>;
```

The above code compiles in gcc et al.  Making the following change allows compilation:

```cpp
  template <typename> struct S
    {
    private:
    struct W
      {
      template <typename> static constexpr auto Impl();
      };

    public:
    template <typename X> using U = decltype(W::template Impl<X>());
    };

  template <typename T> template <typename>
  constexpr auto S<T>::W::Impl()
    { }
```

### PR DESCRIPTION
**Pull request by Fznamznon**
Current context set to where alias was met, not where it is declared caused incorrect access check in case alias referenced private members of the parent class.
This is a recommit of 6b1aa31 with a slight modification in order to fix reported regression.

Fixes https://github.com/llvm/llvm-project/issues/41693

### CODE REVIEW

**Comment by erichkeane** [comment]
Can you point out the diff from the previous patch?

**Comment by Fznamznon** [comment]
> Can you point out the diff from the previous patch?

To fix the following case:
```
// Test case that regressed with the first iteration of the fix
template <typename T> class SP {
    T* data;
};

template <typename T> class A {
    static SP<A> foo();
};

template<typename T> using TRet = SP<A<T>>;

template<typename T> TRet<T> A<T>::foo() { return TRet<T>{};}; // failed with error bc TRet and SP<A>
                                                                                      // had different canonical types due to my change.
```
I had to disable `Sema::CurContext` switch when the type is rebuilt by `Sema::RebuildTypeInCurrentInstantiation` routine since its whole point is to rebuild the type looking at it from different context. There is comment above definition of `Sema::RebuildTypeInCurrentInstantiation` explaining what it does. 

The concrete important diff is:
```
--- a/clang/include/clang/Sema/Sema.h
+++ b/clang/include/clang/Sema/Sema.h
@@ -8738,7 +8738,7 @@ public:
                              SourceLocation IILoc,
                              bool DeducedTSTContext = true);

-
+  bool RebuildingTypesInCurrentInstantiation = false;
   TypeSourceInfo *RebuildTypeInCurrentInstantiation(TypeSourceInfo *T,
                                                     SourceLocation Loc,
                                                     DeclarationName Name);

diff --git a/clang/lib/Sema/SemaTemplate.cpp b/clang/lib/Sema/SemaTemplate.cpp
index 3157700607e0..810a93f37d74 100644
--- a/clang/lib/Sema/SemaTemplate.cpp
+++ b/clang/lib/Sema/SemaTemplate.cpp
@@ -39,6 +39,7 @@
 #include "llvm/ADT/SmallBitVector.h"
 #include "llvm/ADT/SmallString.h"
 #include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/SaveAndRestore.h"

 #include <iterator>
 #include <optional>
@@ -3990,12 +3991,12 @@ QualType Sema::CheckTemplateIdType(TemplateName Name,
     if (Inst.isInvalid())
       return QualType();

-    {
-      bool ForLambdaCallOperator = false;
-      if (const auto *Rec = dyn_cast<CXXRecordDecl>(Pattern->getDeclContext()))
-        ForLambdaCallOperator = Rec->isLambda();
-      Sema::ContextRAII SavedContext(*this, Pattern->getDeclContext(),
-                                     !ForLambdaCallOperator);
+    if (!RebuildingTypesInCurrentInstantiation) {
+      Sema::ContextRAII SavedContext(*this, Pattern->getDeclContext());
+      CanonType =
+          SubstType(Pattern->getUnderlyingType(), TemplateArgLists,
+                    AliasTemplate->getLocation(), AliasTemplate->getDeclName());
+    } else {
       CanonType =
           SubstType(Pattern->getUnderlyingType(), TemplateArgLists,
                     AliasTemplate->getLocation(), AliasTemplate->getDeclName());
@@ -11399,6 +11400,8 @@ TypeSourceInfo *Sema::RebuildTypeInCurrentInstantiation(TypeSourceInfo *T,
   if (!T || !T->getType()->isInstantiationDependentType())
     return T;

+  llvm::SaveAndRestore DisableContextSwitchForTypeAliases(
+      RebuildingTypesInCurrentInstantiation, true);
   CurrentInstantiationRebuilder Rebuilder(*this, Loc, Name);
   return Rebuilder.TransformType(T);
 }

```

File: clang/lib/Sema/SemaTemplate.cpp [hunk 1]
```diff
@@ -3990,9 +3991,16 @@ QualType Sema::CheckTemplateIdType(TemplateName Name,
     if (Inst.isInvalid())
       return QualType();
 
-    CanonType = SubstType(Pattern->getUnderlyingType(),
-                          TemplateArgLists, AliasTemplate->getLocation(),
-                          AliasTemplate->getDeclName());
+    if (!RebuildingTypesInCurrentInstantiation) {
```
Comment (erichkeane) [inline clang/lib/Sema/SemaTemplate.cpp:3994]: This is probably better/less likely to be screwed up in the future if we instead immediately do a 'SavedContext.pop()` if this condition is true, which leaves the SubstType code from being repeated/accidentially diverging.

**Review by erichkeane — APPROVED** [review]

**Comment by rupprecht** [comment]
After this commit, I'm seeing another "out-of-line definition" error for `operator=` on the following reduced snippet:

```c++
class Foo {};

template <class X, class Y = void>
struct Stuff;

template <class Unused>
struct Stuff<Foo, Unused> {
  Stuff& operator=(Stuff&& that);
};

template <class T>
using Alias = Stuff<Foo, T>;

template <class T>
Alias<T>& Alias<T>::operator=(Alias<T>&& that) {
  return *this;
}
```

Results in:

```c++
<source>:15:21: error: out-of-line definition of 'operator=' does not match any declaration in 'Stuff<Foo, type-parameter-0-0>'
   15 | Alias<T>& Alias<T>::operator=(Alias<T>&& that) {
      |                     ^~~~~~~~
```
Is that error expected?

Live conformance view: https://godbolt.org/z/eanhMEevn

I'm also seeing a separate `template argument for template type parameter must be a type; did you forget 'typename'?` error nearby. I could add `typename` to appease the compiler, although I'm not sure if it's supposed to be necessary.

**Comment by Fznamznon** [comment]
@rupprecht , no, this is not expected. Please feel free to revert since it is night in my time zone. Otherwise I will revert tomorrow.

**Comment by rupprecht** [comment]
Thanks to @MaskRay for handling the revert.

A reduction for the second `typename` breakage looks similar, so may be the same underlying issue. But here it is anyway, in case it ends up being a second bug in the patch:

```c++
class Foo {};

template <class T>
struct Ptr {};

template <class X, class Y = void>
class Data;

template <class Unused>
struct Data<Foo, Unused> {
  using Type = int;
  void func();
};

template <class T>
using Alias = Data<Foo, T>;

template <class T>
void Alias<T>::func() {
  auto x = Ptr<Alias<T>::Type>();
  (void)x;
}
```

Results in:

```c++
<source>:20:16: error: template argument for template type parameter must be a type; did you forget 'typename'?
   20 |   auto x = Ptr<Alias<T>::Type>();
      |                ^
      |                typename 
```

Live link, until compiler explorer catches up to the revert: https://godbolt.org/z/zvT4jzqbv

I'd be happy to just add `typename` in this case if the language says it's supposed to be necessary here, but I'd be just as happy to not need to make any code changes :)

**Pull request merged**